fs/fssnapshotfs/snapshot_fs.go

```diff
@@ -0,0 +1,305 @@
+package fssnapshotfs
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+
+	fs_snapshot "github.com/pescuma/go-fs-snapshot/lib"
+
+	"github.com/kopia/kopia/fs"
+	"github.com/kopia/kopia/fs/localfs"
+	"github.com/kopia/kopia/snapshot"
+	"github.com/kopia/kopia/snapshot/policy"
+)
+
+type snapshotEntry struct {
+	policyTree      *policy.Tree
```

Comment:
I think this is unnecessarily coupling this package with policies, while you only need one bit of information.

Can we remove this dependency on policies and lift the decision to use snapshots to the caller?

Reply:
One idea - (which may or may not work) - can we merge this package with `localfs` and explicitly add snapshotting support there (while still keeping it loosely coupled and not dependent on `fs_snapshot` library).

I'm thinking the API could be something like:

```
package localfs

type SnapshotFunc func(originalDir string) (newDir string, closeFunc func(), err error)

func NoSnapshot(originalDir string) (newDir string, closeFunc func(), err error) {
  return oriignalDir, func() {}, nil
}

localfs.NewEntry(path, SnapshotFunc)
```

We would call it by:

```
localfs.NewEntry(path, localfs.NoSnapshot)
localfs.NewEntry(path, fssnapshot.SnapshotFuncFromPolicy(policyTree))
```

This decouples Kopia implementation from specifics of any snapshotting library - anything that matches this function signature can be used as long as it can create snapshots and close them.

Reply:
The reason I put the dependency on `policyTree` is that I wanted to allow the user to configure fs snapshots just for some folders. So it needs to track the policy at each `Child` and `IterateEntries` call, and these do not receive a policy as parameter. I can move it to a function, but I'd need to still store the policy inside there. And it would also always use the function with fs snapshots enabled, because the root folder may have fs snapshots disabled, but a folder inside it may change that. Or maybe I'm over complicating things and we could use only the configuration from the root folder.

About the function, I think it is a good idea to not only depend on fs_snapshots and a function (or something like it) there makes sense. But I think a wrapper around `localfs` is a better approach because it keeps the 2 concerns separated: interact with fs and map different paths.

Reply:
i would not do snapshots within folders as it makes it much harder to reason about things - say you had N policies - consistency-wise you'd still have N snapshots that are not at the same point in time - let's just limit this to the root and keep it simple.

I really don't think we should be depending on policies in this layer.

Reply:
The idea is not to have one snapshot per policy, but to re-use snapshots when possible. The problem with only considering the root snapshot is:
Suppose you ask kopia to snapshot /a and you have in /a/b another drive mounted and in /a/c a different one. The options at that point is (1) to snapshot all mounted drives inside /a or (2) snapshot those as we see them. I went with (2) because /a/b can be ignored by some parameter and we may need not to snapshot it. Because of that, I though it makes sense to also allow to choose that by policies. If you prefer, I can change it to to (1). 

As a side note, this may seen not applicable to Windows, but I discovered that it allows to mount drives at random places too (don't know if anyone uses it). For MacOS (and maybe in the future Linux) that is common concern.